### ISSUE TITLE
dask_array matmul/@ changes the chunk_type

### ISSUE THREAD

**Issue by mughetto** [dask/dask#7126]
**What happened**:
Multiplying two dask arrays with chunk_type=cupy.ndarray results in a dask _array with numpy.ndarray

**What you expected to happen**:
Result should be cupy.ndarray

**Minimal Complete Verifiable Example**:

```python

from dask_cuda import LocalCUDACluster
from dask.distributed import Client
import dask.array as da

import cupy


def RESCAL_score_fn(head_emb, rel_emb, tail_emb):
    print(head_emb)
    print(rel_emb)
    print(tail_emb)
    int_res = da.matmul(head_emb, rel_emb)
    print("types after matmul")
    print(int_res)
    result = (int_res @ tail_emb.T)
    print("types after 2nd matmul")
    print(result)
    return result

if __name__ =='__main__':

    cluster = LocalCUDACluster(CUDA_VISIBLE_DEVICES='1,2,3')
    client = Client(cluster)

    rs = da.random.RandomState(RandomState=cupy.random.RandomState)
    my_arr = rs.random((10000, 10000), chunks=(1000, 1000))

    embeddings = RESCAL_score_fn(my_arr,my_arr,my_arr)
    print(embeddings)
```

**Environment**:

- Dask version: 2021.01.1
- Python version: 3.7.9
- Operating System: CentOS7, 3.10.0-957.27.2.el7.x86_64
- Install method (conda, pip, source): conda (RAPIDS 0.17)

**Comment by pentschev** [comment]
Thanks @mughetto for filing the issue, I can reproduce and it seems related to https://github.com/dask/dask/blob/9bb586a6b8fac1983b7cea3ab399719f93dbbb29/dask/array/tests/test_cupy.py#L61-L72 .

**Comment by pentschev** [comment]
Investigating a bit further, this is also blocked by https://github.com/dask/dask/pull/6738 . There's still a lot of work to be done, including lots of functions from https://github.com/dask/dask/blob/master/dask/array/routines.py . I plan to get back to that but only after NumPy 1.20 is released, as it will be a minimum requirement for such functions to work.

**Comment by mughetto** [comment]
Ok, I understand, thanks a lot!

In the meantime do you have any suggestion of how to practically go around this and stay on GPU? Maybe directly numba ?

**Comment by pentschev** [comment]
If you want to follow the Numba path with Dask, [this article](https://blog.dask.org/2019/06/19/python-gpus-status-update) may be helpful, but please note this may be a bit outdated. In that case, of course, you need to write your own [Numba matrix multiplication](https://numba.pydata.org/numba-doc/dev/cuda/examples.html#matrix-multiplication) kernels. I'm afraid there are no simpler alternatives today, but hopefully in the next few months we'll have NumPy released and then #6738 completed and merged into Dask as well.

**Comment by mughetto** [comment]
Hey @pentschev came back to this yesterday and noticed that #6738 was merged. 
I've tested with the most recent of RAPIDS (21.06) that uses numpy 1.21.1 and dask 2021.0.5.0. Now the minimal example above returns:
```
types after 2nd matmul
dask.array<reshape, shape=(10000, 10000), dtype=float64, chunksize=(1000, 1000), chunktype=cupy.ndarray>
dask.array<reshape, shape=(10000, 10000), dtype=float64, chunksize=(1000, 1000), chunktype=cupy.ndarray>
```
So it looks to me like we can close this ticket! What do you think?

**Comment by pentschev** [comment]
Hi @mughetto , sorry I forgot to update this issue, but you're correct, it now works and can be closed. Thanks for pinging here!
